File: Packages/PTCore/Sources/PTCore/Models/Book.swift
import Foundation
import GRDB

public struct Book: Codable, FetchableRecord, PersistableRecord, Identifiable, Sendable, Equatable {
    public static let databaseTableName = "tb_books"

    public var id: Int64?
    public var title: String
    public var coverPath: String
    public var filePath: String
    public var lastReadPosition: String
    public var readingPercentage: Double
    public var author: String
    public var isDeleted: Bool
    public var description: String?
    public var rating: Double
    public var groupId: Int64
    public var md5: String?
    public var createTime: Date
    public var updateTime: Date

    enum CodingKeys: String, CodingKey {
        case id
        case title
        case coverPath = "cover_path"
        case filePath = "file_path"
        case lastReadPosition = "last_read_position"
        case readingPercentage = "reading_percentage"
        case author
        case isDeleted = "is_deleted"
        case description
        case rating
        case groupId = "group_id"
        case md5 = "file_md5"
        case createTime = "create_time"
        case updateTime = "update_time"
    }

    public mutating func didInsert(_ inserted: InsertionSuccess) {
        id = inserted.rowID
    }

    public static func placeholder(title: String, filePath: String) -> Book {
        let now = Date()
        return Book(
            id: nil,
            title: title,
            coverPath: "",
            filePath: filePath,
            lastReadPosition: "",
            readingPercentage: 0,
            author: "",
            isDeleted: false,
            description: nil,
            rating: 0,
            groupId: 0,
            md5: nil,
            createTime: now,
            updateTime: now
        )
    }
}

File: lib/models/book.dart
import 'dart:typed_data';

import 'package:papertok_reader/utils/get_path/get_base_path.dart';

class Book {
  int id;
  String title;
  String coverPath;
  String filePath;
  String lastReadPosition;
  double readingPercentage;
  String author;
  bool isDeleted;
  String? description;
  double rating;
  int groupId;
  String? md5;
  Uint8List? bookmarkData;
  String sourceKind;
  DateTime createTime;
  DateTime updateTime;

  Book(
      {required this.id,
      required this.title,
      required this.coverPath,
      required this.filePath,
      required this.lastReadPosition,
      required this.readingPercentage,
      required this.author,
      required this.isDeleted,
      this.description,
      required this.rating,
      this.groupId = 0,
      this.md5,
      this.bookmarkData,
      this.sourceKind = 'imported',
      required this.createTime,
      required this.updateTime});

  factory Book.mock() {
    return Book(
      id: 1,
      title: 'Mock Book',
      coverPath: '',
      filePath: '',
      lastReadPosition: '',
      readingPercentage: 0.78,
      author: 'PaperTok Reader',
      isDeleted: false,
      rating: 0,
      bookmarkData: null,
      sourceKind: 'imported',
      createTime: DateTime.now(),
      updateTime: DateTime.now(),
    );
  }

  String get coverFullPath {
    return getBasePath(coverPath);
  }

  String get fileFullPath {
    return getBasePath(filePath);
  }

  bool get isInPlace => sourceKind == 'inplace';

  Map<String, Object?> toMap() {
    return {
      'title': title,
      'cover_path': coverPath,
      'file_path': filePath,
      'last_read_position': lastReadPosition,
      'reading_percentage': readingPercentage,
      'author': author,
      'is_deleted': isDeleted ? 1 : 0,
      'description': description,
      'rating': rating,
      'group_id': groupId,
      'file_md5': md5,
      'bookmark_data': bookmarkData,
      'source_kind': sourceKind,
      'create_time': createTime.toIso8601String(),
      'update_time': updateTime.toIso8601String(),
    };
  }

  Book copyWith({
    int? id,
    String? title,
    String? coverPath,
    String? filePath,
    String? lastReadPosition,
    double? readingPercentage,
    String? author,
    bool? isDeleted,
    String? description,
    double? rating,
    int? groupId,
    String? md5,
    Uint8List? bookmarkData,
    String? sourceKind,
    DateTime? createTime,
    DateTime? updateTime,
  }) {
    return Book(
      id: id ?? this.id,
      title: title ?? this.title,
      coverPath: coverPath ?? this.coverPath,
      filePath: filePath ?? this.filePath,
      lastReadPosition: lastReadPosition ?? this.lastReadPosition,
      readingPercentage: readingPercentage ?? this.readingPercentage,
      author: author ?? this.author,
      isDeleted: isDeleted ?? this.isDeleted,
      description: description ?? this.description,
      rating: rating ?? this.rating,
      groupId: groupId ?? this.groupId,
      md5: md5 ?? this.md5,
      bookmarkData: bookmarkData ?? this.bookmarkData,
      sourceKind: sourceKind ?? this.sourceKind,
      createTime: createTime ?? this.createTime,
      updateTime: updateTime ?? this.updateTime,
    );
  }

  factory Book.fromDb(Map<String, dynamic> map) {
    return Book(
      id: map['id'] as int,
      title: map['title'] as String? ?? '',
      coverPath: map['cover_path'] as String? ?? '',
      filePath: map['file_path'] as String? ?? '',
      lastReadPosition: map['last_read_position'] as String? ?? '',
      readingPercentage: (map['reading_percentage'] as num?)?.toDouble() ?? 0.0,
      author: map['author'] as String? ?? '',
      isDeleted: (map['is_deleted'] as int? ?? 0) == 1,
      description: map['description'] as String?,
      rating: (map['rating'] as num?)?.toDouble() ?? 0.0,
      groupId: map['group_id'] as int? ?? 0,
      md5: map['file_md5'] as String?,
      bookmarkData: map['bookmark_data'] as Uint8List?,
      sourceKind: map['source_kind'] as String? ?? 'imported',
      createTime: DateTime.parse(map['create_time'] as String),
      updateTime: DateTime.parse(map['update_time'] as String),
    );
  }
}
